File: cane-csharp/cane-core/Runtime/Constraints/PathProjection.cs
using System;

namespace Cane.Constraints
{
    internal sealed partial class PathSampler
    {
        // The inverse table has 24 subdivisions. Forward sampling retains its independent ten-step lookup.
        internal float? ProjectDistance(Point target)
        {
            double bestSquared = double.PositiveInfinity, bestDistance = 0;
            bool found = false;
            void Consider(double x, double y, double distance)
            {
                double dx = target.X - x, dy = target.Y - y, squared = dx * dx + dy * dy;
                if (double.IsNaN(distance) || double.IsInfinity(distance) || !(squared < bestSquared)) return;
                bestSquared = squared; bestDistance = distance; found = true;
            }
            void Segment(Point from, Point to, double startDistance, double endDistance)
            {
                double dx = (double)to.X - from.X, dy = (double)to.Y - from.Y, squared = dx * dx + dy * dy;
                if (!(squared > 0) || double.IsInfinity(squared)) return;
                double ratio = Math.Max(0, Math.Min(1, (((double)target.X - from.X) * dx + ((double)target.Y - from.Y) * dy) / squared));
                Consider(from.X + dx * ratio, from.Y + dy * ratio, startDistance + (endDistance - startDistance) * ratio);
            }
            for (int i = 0; i < curves.Length; i++)
            {
                Curve curve = curves[i]; double begin = i == 0 ? 0 : lengths[i - 1], span = lengths[i] - begin;
                float[]? localLengths = detailed == null ? null : curve.Lengths(24);
                double localTotal = localLengths == null ? 0 : localLengths[23];
                Point previous = curve.P0; double before = begin;
                for (int subdivision = 0; subdivision < 24; subdivision++)
                {
                    Point next = curve.Point((subdivision + 1) / 24f);
                    double ratio = localLengths == null ? (subdivision + 1) / 24.0 : localTotal > 0 ? localLengths[subdivision] / localTotal : 0;
                    double after = begin + span * ratio;
                    Segment(previous, next, before, after); previous = next; before = after;
                }
            }
            if (!closed && curves.Length != 0)
            {
                void Ray(Point anchor, Point handle, bool start)
                {
                    double dx = start ? (double)handle.X - anchor.X : (double)anchor.X - handle.X;
                    double dy = start ? (double)handle.Y - anchor.Y : (double)anchor.Y - handle.Y;
                    double length = Math.Sqrt(dx * dx + dy * dy);
                    if (!(length > 0) || double.IsInfinity(length)) return;
                    dx /= length; dy /= length;
                    double distance = ((double)target.X - anchor.X) * dx + ((double)target.Y - anchor.Y) * dy;
                    if (start ? distance >= 0 : distance <= 0) return;
                    Consider(anchor.X + dx * distance, anchor.Y + dy * distance, start ? distance : Length + distance);
                }
                Ray(curves[0].P0, curves[0].P1, true);
                Ray(curves[curves.Length - 1].P3, curves[curves.Length - 1].P2, false);
            }
            return found ? (float)bestDistance : (float?)null;
        }
    }
}

File: cane-cpp/cane-core/src/constraints/path_projection.cpp
#include "path_sampler.hpp"
#include <algorithm>
#include <cmath>
#include <limits>

namespace cane::constraints {
std::optional<float> PathSampler::project_distance(Point target) const noexcept {
    if (!available() || !std::isfinite(target.x) || !std::isfinite(target.y)) return {};
    auto best_squared = std::numeric_limits<float>::infinity(); std::optional<float> best;
    const auto consider = [&](Point point, float distance) {
        const auto dx = target.x - point.x, dy = target.y - point.y, squared = dx * dx + dy * dy;
        if (std::isfinite(distance) && squared < best_squared) { best_squared = squared; best = distance; }
    };
    const auto segment = [&](Point from, Point to, float begin, float end) {
        const auto dx = to.x - from.x, dy = to.y - from.y, squared = dx * dx + dy * dy;
        if (squared == 0 || !std::isfinite(squared)) return;
        const auto ratio = std::clamp(((target.x - from.x) * dx + (target.y - from.y) * dy) / squared, 0.0f, 1.0f);
        consider({from.x + dx * ratio, from.y + dy * ratio}, begin + (end - begin) * ratio);
    };
    float begin = 0;
    for (const auto& c : curves_) {
        std::array<float, 24> local{}; const auto local_total = constant_speed_ ? c.lengths(24, local.data()) : 0;
        const auto span = c.cumulative - begin; auto previous = c.p0; auto before = begin;
        for (unsigned i = 0; i < 24; ++i) {
            const auto t = static_cast<float>(i + 1) / 24; const auto next = c.point(t);
            const auto progress = constant_speed_ ? local_total > 0 ? local[i] / local_total : 0 : t;
            const auto after = begin + span * progress;
            segment(previous, next, before, after); previous = next; before = after;
        }
        begin = c.cumulative;
    }
    if (!closed_) {
        const auto ray = [&](Point anchor, Point handle, bool start) {
            auto dx = start ? handle.x - anchor.x : anchor.x - handle.x;
            auto dy = start ? handle.y - anchor.y : anchor.y - handle.y;
            const auto magnitude = cane::hypot(dx, dy); if (magnitude == 0 || !std::isfinite(magnitude)) return;
            dx /= magnitude; dy /= magnitude;
            const auto distance = (target.x - anchor.x) * dx + (target.y - anchor.y) * dy;
            if (start ? distance >= 0 : distance <= 0) return;
            consider({anchor.x + dx * distance, anchor.y + dy * distance}, start ? distance : length() + distance);
        };
        ray(curves_.front().p0, curves_.front().p1, true); ray(curves_.back().p3, curves_.back().p2, false);
    }
    return best;
}
}
